### PR TITLE
[bluetooth] Add support for devices jumping between adapters.

### LINKED ISSUE #7008: [bluetooth] Allow bluetooth devices to jump between adapters.
This is a feature that is available on the unofficial bluetooth binding but not the official one.
It can be implemented by having a new type of BluetoothAdapter that produces "roaming" BluetoothDevices, which would basically delegate all their functionality to devices from other BluetoothAdapters.

In order to make these roaming devices more useful they should add two dynamic channels:
1. ThingUID of currently used adapter (the one it is connected to, or closest one if not)
2. Location of currently used adapter

### PR DESCRIPTION
**Pull request by cpmeister**
Implements and closes #7008 

This PR basically adds support for a virtual bluetooth adapter (what I call a *roaming* adapter) that allows handlers to interact with a virtual BluetoothDevice that delegates all its operations to actual BluetoothDevice instances on actual adapters. The virtual adapter will always delegate to the nearest actual adapter for a given bluetooth device.

The primary purpose of this new feature is presence detection. If a user places bluetooth adapters all of the house they can identify which room a particular bluetooth device is in at any time.

This is one of the features that I'm trying to replicate from the 3rd party bluetooth binding.

### CODE REVIEW

**Comment by cpmeister** [comment]
@openhab/add-ons-maintainers I feel this is ready to be merged.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/resources/ESH-INF/thing/airthings.xml [hunk 1]
```diff
@@ -14,8 +15,6 @@
 		<description>Indoor air quality monitor with radon detection</description>
 
 		<channels>
-			<channel id="rssi" typeId="rssi"/>
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/resources/ESH-INF/thing/airthings.xml:17]: Does it harm to have it listed here? If Things are created manually, this has the advantage that this channel is available without any handler required to be present.

File: bundles/org.openhab.binding.bluetooth/README.md [hunk 2]
```diff
@@ -19,10 +19,11 @@ This should be the best choice for any Linux-based single board computers like e
...
 
-| Thing Type ID | Description                                                                                             |
-|---------------|---------------------------------------------------------------------------------------------------------|
-| beacon        | A Bluetooth device that is not connected, but only broadcasts annoucements.                             |
-| connected     | A Bluetooth device that allows a direct connection and which provides specific services when connected. |
+| Thing Type ID | Description                                                                                                |
+|---------------|------------------------------------------------------------------------------------------------------------|
+| beacon        | A Bluetooth device that is not connected, but only broadcasts annoucements.                                |
+| connected     | A Bluetooth device that allows a direct connection and which provides specific services when connected.    |
+| roaming       | A virtual Bluetooth adapter that interacts with Bluetooth devices through their nearest Bluetooth adapter. |
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth/README.md:26]: This is not the right place to add it - if I get it right, "roaming" is a new adapter (next to BlueZ and BlueGiga) and not a BT device type (beacon or connected).

File: bundles/org.openhab.binding.bluetooth/README.md [hunk 3]
```diff
@@ -37,11 +38,12 @@ Both thing types only require a single configuration parameter `address`, which
...
-Every Bluetooth thing has the following channel:
+Every Bluetooth thing has the following channels:
 
-| Channel ID | Item Type | Description                                                                                         |
-|------------|-----------|-----------------------------------------------------------------------------------------------------|
-| rssi       | Number    | The "Received Signal Strength Indicator", the [RSSI](https://blog.bluetooth.com/proximity-and-rssi) |
+| Channel ID    | Item Type | Description                                                                                         |
+|---------------|-----------|-----------------------------------------------------------------------------------------------------|
+| rssi          | Number    | The "Received Signal Strength Indicator", the [RSSI](https://blog.bluetooth.com/proximity-and-rssi) |
+| last-activity | DateTime  | The last time that any radio activity was received from this device                                 |
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth/README.md:46]: How is that related to this PR?
I am not too keen on such channels - if we feel such information is needed, we should rather solve it for ALL bindings at the same time.
What I could live with here is to keep track of it as a Thing property instead.

File: bundles/org.openhab.binding.bluetooth/README.md [hunk 4]
```diff
@@ -50,18 +52,31 @@ Every Bluetooth thing has the following channel:
...
 
+Devices which use a `roaming` adapter as their bridge also gain the following channels:
+
+| Channel ID       | Item Type | Description                                                     |
+|------------------|-----------|-----------------------------------------------------------------|
+| adapter-uid      | String    | The thingUID of the adapter that is nearest to this device      |
+| adapter-location | String    | The value of the `Location` specified for the nearest adapter    |
+
+
+## Roaming Adapter
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth/README.md:63]: Just like the other adapters, this one should go into its own bundle. Is there any strong reason against that?

File: bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml [hunk 5]
```diff
@@ -5,11 +5,32 @@
 	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
 
 	<channel-type id="rssi">
-		<item-type>Number</item-type>
+		<item-type>Number:Power</item-type>
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml:8]: Are you sure? According to https://en.wikipedia.org/wiki/DBm, "dBm" is dimensionless.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/resources/ESH-INF/thing/airthings.xml (hunk 1 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/resources/ESH-INF/thing/airthings.xml:17]: I didn't realize they had that use case. I'll add them back in then.

File: bundles/org.openhab.binding.bluetooth/README.md (hunk 2 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth/README.md:26]: You are correct, "roaming" is a type of adapter, not a regular bluetooth device. I just put it in this table since I thought all thing types should be listed there. Should I create a separate section/table for the roaming adapter?

File: bundles/org.openhab.binding.bluetooth/README.md (hunk 3 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth/README.md:46]: This PR is a general adaptation of the features available in the 3rd party bluetooth binding. One of the primary features of the binding was presence detection, which a `last-activity` channel would almost certainly be required for. 

While beacon bluetooth devices could be tracked by a change in their RSSI channel, such a method can't be used for connected devices which don't send out advertisement information that would trigger an RSSI channel update. (This is the case for devices tracked through Bluez adapters, but not for Bluegiga, because they are special I guess).

The "last-activity" channel would be updated on either rssi changes or on successful responses during connected communications.

I can change this to a thing property if you wish, but I don't know if users will be able to make rules that would use it then. WDYT?

File: bundles/org.openhab.binding.bluetooth/README.md (hunk 4 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth/README.md:63]: I initially attempted something akin to that approach but I found that the BluetoothDiscoveryService would need to handle roaming bluetooth devices differently than other adapters, and in order to do that they would need awareness of what a roaming was. This awareness, as well as an overall increase in the complexity of the problems that would need to be resolved, made me decide to just embed the roaming adapter into the base bluetooth bundle itself.

This does make sense in a way, since the roaming adapter isn't any sort of concrete implementation of a Bluetooth adapter as it delegates all of its operations to a real adapter instead.

File: bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml (hunk 5 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml:8]: I just looked up the smarthome unit that used "dBm" as its symbol. `SmartHomeUnits.DECIBEL_MILLIWATTS` uses "Power" as its quantity type.

Also I think that the wiki article states that "dB" is dimensionless, not "dBm".

**Comment by cpmeister** [comment]
@kaikreuzer Can you answer some of the questions I posted?

File: bundles/org.openhab.binding.bluetooth/README.md (hunk 2 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth/README.md:26]: As mentioned further below, this should ideally be a separate bundle, just like the other adapters.

File: bundles/org.openhab.binding.bluetooth/README.md (hunk 3 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth/README.md:46]: Can you elaborate what kind of rules people write when having that channel available? Could the logic of such rules possibly be captured within the binding and be provided as a special channel?

For presence detection, I would possibly consider the Thing status to be the best instrument - if the Thing is OFFLINE, it means the device isn't in reach anymore. BT things could have a configuration parameter, which says after what period of inactivity the Thing should move to OFFLINE status. Would that address the requirement?

File: bundles/org.openhab.binding.bluetooth/README.md (hunk 4 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth/README.md:63]: The roaming adapter is nonetheless just an optional feature, which could be left away easily and not a mandatory core part of the binding. I think wrt code cleanliness and structure, it would be nice to have it as a separate bundle.
Wrt the discovery service I'd suggest to think about a way to introduce an interface to the "generic" discovery service, which allows the customization of its logic by extensions. Similar to the "UPnPDiscoveryParticipant", which plugs into the "UPnPDiscoveryService".
Do you think that would be feasible?

**Comment by kaikreuzer** [comment]
> Can you answer some of the questions I posted?

Sure, sorry for the delay. Was sick a few days and had a few other urgent tasks to finish. Time is scarce these days for me :-/.

File: bundles/org.openhab.binding.bluetooth/README.md (hunk 3 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth/README.md:46]: > For presence detection, I would possibly consider the Thing status to be the best instrument - if the Thing is OFFLINE, it means the device isn't in reach anymore. BT things could have a configuration parameter, which says after what period of inactivity the Thing should move to OFFLINE status. Would that address the requirement?

You have a point there. I'll adjust the feature to control thing status and get rid of the `last-activity` channel. Thank for pointing this out.

File: bundles/org.openhab.binding.bluetooth/README.md (hunk 4 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth/README.md:63]: > Wrt the discovery service I'd suggest to think about a way to introduce an interface to the "generic" discovery service, which allows the customization of its logic by extensions. Similar to the "UPnPDiscoveryParticipant", which plugs into the "UPnPDiscoveryService".
Do you think that would be feasible?

I think that should be feasible, I'll get to work on that. Thanks!

**Comment by cpmeister** [comment]
> Sure, sorry for the delay. Was sick a few days and had a few other urgent tasks to finish. Time is scarce these days for me :-/.

Don't worry about it. Thanks for getting back to me. 
Stay safe and I hope you are feeling better.

**Comment by cpmeister** [comment]
I went ahead and removed the `last-activity` channel since I'll add the idle offline feature in another PR.

File: bundles/org.openhab.binding.bluetooth.roaming/README.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+| roaming        | A virtual Bluetooth adapter that interacts with Bluetooth devices through their nearest Bluetooth adapter. |
+
+## Channels
+
+Devices which use a `roaming` adapter as their bridge also gain the following channels:
+
+| Channel ID       | Item Type | Description                                                     |
+|------------------|-----------|-----------------------------------------------------------------|
+| adapter-uid      | String    | The thingUID of the adapter that is nearest to this device      |
+| adapter-location | String    | The value of the `Location` specified for the nearest adapter    |
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth.roaming/README.md:20]: To me as a reader it is completely unclear what this "Location" string will contain. Would be nice to have some better explanation for it.

File: bundles/org.openhab.binding.bluetooth.roaming/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+## Discovery
+
+A default roaming adapter is discovered automatically, although you can still configure it textually if you want.
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth.roaming/README.md:24]: Why? Discovery is meant for real stuff where it is clear that it exists. The roaming adapter is a feature that the user can decide by himself whether he wants to make use of it or not. If he does, he can simply add a roaming bridge very easily.

File: bundles/org.openhab.binding.bluetooth.roaming/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+There can be only a single roaming adapter on a system.
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth.roaming/README.md:25]: If this is due to technical constraints, it smells like some unclean design. In general, it should be allowed to create any number of things/bridges of a given type.

File: bundles/org.openhab.binding.bluetooth.roaming/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+## Bridge Configuration
+
+The roaming bridge requires an `address` parameter which mearly servers as an identifier for the virtual adapter, its value doesn't really matter.
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth.roaming/README.md:29]:
```suggestion
The roaming bridge requires an `address` parameter which merely serves as an identifier for the virtual adapter, its value does not matter. 
```

File: bundles/org.openhab.binding.bluetooth.roaming/README.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+## Discovery
+
+A default roaming adapter is discovered automatically, although you can still configure it textually if you want.
+There can be only a single roaming adapter on a system.
+
+## Bridge Configuration
+
+The roaming bridge requires an `address` parameter which mearly servers as an identifier for the virtual adapter, its value doesn't really matter. 
+It is advised to make sure it is set to a value outside of the normal bluetooth mac address range, for example: "FF:FF:FF:FF:FF:FF".
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth.roaming/README.md:30]: This sounds as if the value should be a mac address. Could it also be `/(%§&@~$")` or is its syntax checked against mac structure?

File: bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml (hunk 5 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml:8]: We might have to fix SmartHomeUnits then...

> Also I think that the wiki article states that "dB" is dimensionless, not "dBm".

I guess you are just reading the wrong sentence in it. It states:
"The dBm is also dimensionless but since it compares to a fixed reference value the dBm rating is an absolute one."

File: features/openhab-addons/src/main/resources/footer.xml [hunk 11]
```diff
@@ -1,4 +1,4 @@
-    <!-- aggregated features -->
+	<!-- aggregated features -->
```
Comment (kaikreuzer) [inline features/openhab-addons/src/main/resources/footer.xml:1]: please revert this change - the indentation was correct before

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 8 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:25]: Well I was worried a bit of how that particular aspect would blow over. Technically I could support more than one but since every roaming adapter monitors every other real adapter I didn't see a reason to allow supporting more than one since they would both report the same thing. So the roaming adapter limitation was more a design choice rather than a technical limitation. I can add support for it if you think that is what I should do.

I suppose down the line multiple roaming adapters could come in handy if I added support for manually choosing which real adapters a roaming adapter would monitor. Then you could have multiple roaming adapters each responsible for a separate set of real adapters.

But I don't know if that is something worth pursuing at the moment since the number of bluetooth adapters available to openhab is limited to the number that can be directly connected to the device that openhab is running on (might not be the case for bluegiga adapters if you use serial->net software).
But once ble->mqtt support comes down the line then adding that feature would very likely need to be considered.

WDYT?

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 7 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:24]: I'll remove it then since users could always create it manually if they wanted.

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 8 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:25]: If it is not so much a technical constraint, but rather a functional/logical one, then this sounds good to me. I mainly wanted to make sure that such "logic simplifications" do not result in shortcuts in the code, which would go against the architectural guidelines (to have things independent from each other and individually referenceable).
No need to make any use of the fact that there might be more than one through features in the code right now.

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 10 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:30]: Right now the value is required to be a mac address to create an instance of `BluetoothAddress` for the adapter's `getAddress()` method to return.

Funnily enough, after looking through the Bluetooth binding code it seems that only place that BluetoothAdapter.getAddress() is even called is in a logging statement. So there isn't really any architectural requirement for `BluetoothAdapter` to even have a getAddress method.
Would it make sense to remove that method from the `BluetoothAdapter` interface to allow users to create roaming thing instances without having to specify a fake bluetooth mac address?

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 6 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:20]: Will do

File: bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml (hunk 5 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml:8]: > "The dBm is also dimensionless but since it compares to a fixed reference value the dBm rating is an absolute one."

Huh, does this mean that I shouldn't be using `SmartHomeUnits.DECIBEL_MILLIWATTS` then since it is wrongly typed as a unit of Power? Should probably be fixed then.

I'll leave rssi as a regular Number type for now then until the units get sorted out. 
Sorry for the back and forth, I should have taken the time to read sentence following the one I posted.

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 10 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:30]: Well, this parameter is what makes the bridge unique - it is the only differentiator between two bridges. Just as you say "it merely serves as an identifier". But this means we cannot remove it (or if we did, we would need to add some other parameter as a replacement). So I guess we can keep it and should only specify, what the value of the address should be for roaming adapters. If it does not have to be a mac, you could come up with something yourself, like saying it has to be `virtual:XX` with XX a byte value or whatever. And having a small syntax check on the parameter then probably won't hurt.

File: bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml (hunk 5 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth/src/main/resources/ESH-INF/thing/channels.xml:8]: > Should probably be fixed then.

Wanna do a PR against openhab-core?

> I'll leave rssi as a regular Number type for now then until the units get sorted out.

Sounds like a reasonable solution for now.

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 10 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:30]: > If it does not have to be a mac, you could come up with something yourself, like saying it has to be virtual:XX with XX a byte value or whatever. And having a small syntax check on the parameter then probably won't hurt.

Well right now it has to be a mac in order to be able to implement the `BluetoothAdapter.getAddress()` method but if that method were to be removed you are correct that the identifier could be anything we want.
But it might also be possible to not have any identifier for roaming bridges at all outside of their thingUID. Roaming bridges only exist as ThingHandlers and have no actual backing device; they exist as soon as a roaming thing requests a handler.

Is the ThingUID not a sufficient bridge identifier? Or do you think a separate identifier is required?

Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:30]: > right now it has to be a mac in order to be able to implement the BluetoothAdapter.getAddress()

But what you are saying is that this method isn't really used anywhere. So why does it have to be a mac when it just says "address"? My idea was to define "address" a bit wider than being a mac and thus use something artificial like `virtual:01` as an address for the roaming bridge.

> Is the ThingUID not a sufficient bridge identifier?

No, as mentioned above, it would mean that it is the very same "thing", just being referenced with different ids. For the roaming bridge it is difficult to argue, since it is an exceptional abstract case. I somehow feel that we nonetheless should not make any exceptions for it, at least when it isn't a big deal for us.

Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:30]: > So why does it have to be a mac when it just says "address"?

It is meant to be reflective of the how the bluetooth protocols are meant to operate. All devices on a bluetooth network have a mac address, host and devices, through which they identify themselves as they communicate across the network. The `BluetoothAdapter.getAddress` method was meant to reflect an attribute of real physical bluetooth adapters which have unique mac addresses.

All `BluetoothAdapter` implementations are required to return a non-null `BluetoothAddress` object from their `BluetoothAdapter.getAddress` implementation.

Because the roaming adapter has to implement `BluetoothAdapter` then it must also be able to provide an BluetoothAddress instance to return. The `BluetoothAddress` enforces the mac format of the address on creation, so their is no way to not have a BluetoothAddress that isn't in the XX:XX:XX:XX:XX:XX format.

So if we want the roaming adapter to support non-mac addresses then we would either have to remove `BluetoothAdapter.getAddress`, change `BluetoothAdapter.getAddress`'s return type to be nullable so that the roaming adapter can return null, change the return type to something that isn't `BluetoothAddress`, or modify `BluetoothAddress` to allow non-mac address formats.
Which one do you think I should go with?

**Comment by cpmeister** [comment]
@kaikreuzer I'd really like to get this merged for 2.5.6 since I have another bluetooth PR that finally actually adds dynamic gatt characteristic channel support. I'd hate to have both of these features pushed back another month.

**Comment by kaikreuzer** [comment]
@cpmeister Sorry that I didn't make it for 2.5.6, but it isn't forgotten. Hope to soon find time to review!

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 10 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:30]: I just read that devices are often using random MAC addresses for privacy reasons. It therefore seems to me that it is legit to just provide ANY MAC address - so your suggestion in the readme to just define any MAC address as long as it is in the valid format might indeed be the simplest way to go?
You say "a value outside of the normal bluetooth mac address range" - can this be specified in more detail?

File: features/openhab-addons/src/main/resources/footer.xml [hunk 12]
```diff
@@ -1,4 +1,4 @@
-	<!-- aggregated features -->
+    <!-- aggregated features -->
```
Comment (kaikreuzer) [inline features/openhab-addons/src/main/resources/footer.xml]:
```suggestion
<!-- aggregated features -->
```

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 10 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:30]: @cpmeister If you want to get it into 2.5.7, we should now finish it soon.
Would you agree with my comment above? If so, please adapt the code, rebase it and we should be good to merge!

**Comment by cpmeister** [comment]
@kaikreuzer I've resolved the conflicts, any chance of this getting merged for 2.5.9? 
Is there anything I'm missing?

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 10 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md:30]: > You say "a value outside of the normal bluetooth mac address range" - can this be specified in more detail?

You give one example now ("FF:FF:FF:FF:FF:FF"), but "value outside normal range" sounds like there should be a kind of "reserved" address space for local/private addresses. What values other than "FF:FF:FF:FF:FF:FF" would you think are ok to use?

File: features/openhab-addons/src/main/resources/footer.xml [hunk 13]
```diff
@@ -1,4 +1,113 @@
 	<!-- aggregated features -->
+	<feature name="openhab-binding-bluetooth" description="Bluetooth Binding" version="${project.version}">
```
Comment (kaikreuzer) [inline features/openhab-addons/src/main/resources/footer.xml:2]: Here seems to be something wrong, you're duplicating all information in this file.

File: bundles/pom.xml [hunk 14]
```diff
@@ -1,6 +1,5 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (kaikreuzer) [inline bundles/pom.xml:2]: Spotless complains about this file, so you should apply it here as well, I guess.

**Review by kaikreuzer — CHANGES_REQUESTED** [review]

**Comment by kaikreuzer** [comment]
@cpmeister Please do not do any merge commits on PRs - always use "rebase" instead, thanks!

File: bundles/pom.xml (hunk 14 above)
Comment (cpmeister) [reply bundles/pom.xml:2]: strange, spotless was the one that made that change... I'll manually edit it back

File: features/openhab-addons/src/main/resources/footer.xml (hunk 13 above)
Comment (cpmeister) [reply features/openhab-addons/src/main/resources/footer.xml:2]: Something wrong indeed, I'll fix it.

**Comment by cpmeister** [comment]
@kaikreuzer I tried rebasing but it ended producing 20+ conflicting files and I couldn't make heads or tails of which version of the conflicted files were correct. I keep trying but I've never once had a successful rebase when conflicts were involved.
With a merge I only had to deal with ~5 conflicting files, much more manageable, although in this case I can still make mistakes when resolving conflicts with large formatting changes. 🤷

File: features/openhab-addons/src/main/resources/footer.xml (hunk 13 above)
Comment (cpmeister) [reply features/openhab-addons/src/main/resources/footer.xml:2]: Strange, I've fixed it on my branch but it doesn't show it as fixed here...

Comment (cpmeister) [reply features/openhab-addons/src/main/resources/footer.xml]: nvm, fixed it

File: bundles/org.openhab.binding.bluetooth.roaming/README.md (hunk 10 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md]: I went ahead and just removed the address from the roaming adapters entirely. It wasn't meaningfully used in the code and would only confuse users trying to configure a roaming adapter thing.

Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md]: Hm, it seems you forgot about all previous discussion we had here - removing the address was not really an option, see https://github.com/openhab/openhab-addons/pull/7199#discussion_r439770589.

Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md]: > Hm, it seems you forgot about all previous discussion we had here - removing the address was not really an option, see #7199 (comment).

Apparently I did forget! But since I loathe idea of a useless configuration parameter I went ahead and implemented the roaming adapter groups feature. This feature adds a `groupUIDs` parameter which could serve as a thing identifier for any roaming bridge thing. I've also eliminated the restriction of only allowing a single roaming adapter thing to be defined.

I hope this is sufficient compromise.

Comment (kaikreuzer) [reply bundles/org.openhab.binding.bluetooth.roaming/README.md]: Ok, sounds like a decent compromise, thank you!

File: bundles/org.openhab.binding.bluetooth.roaming/README.md [hunk 15]
```diff
@@ -22,11 +22,11 @@ Devices which use a `roaming` adapter as their bridge also gain the following ch
 ## Discovery
 
 Roaming adapters cannot be discovered, they can only be created manually.
-There can be only a single roaming adapter on a system.
 
 ## Bridge Configuration
+The Roaming bridge has an optional parameter `groupUIDs` that configures which Bluetooth adapters this roaming bridge will be monitored for the purpose of roaming devices. `groupUIDs` must be formatted as a comma separated list of Bluetooth adapter thing UID values. If the `groupUIDs` parameter is not specified or left empty then the Roaming adapter will track devices across all other Bluetooth adapters.
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.bluetooth.roaming/README.md:27]:
```suggestion

The Roaming bridge has an optional parameter `groupUIDs` that configures which Bluetooth adapters this roaming bridge will be monitored for the purpose of roaming devices.
`groupUIDs` must be formatted as a comma separated list of Bluetooth adapter thing UID values.
If the `groupUIDs` parameter is not specified or left empty then the Roaming adapter will track devices across all other Bluetooth adapters.
```

**Review by kaikreuzer — APPROVED** [review]
